clients/roscpp/src/libros/service_server_link.cpp

```diff
@@ -362,15 +368,37 @@ bool ServiceServerLink::call(const SerializedMessage& req, SerializedMessage& re
     processNextCall();
   }
 
+  bool timed_out = false;
   {
     boost::mutex::scoped_lock lock(info->finished_mutex_);
 
+    using namespace boost::chrono;
+    steady_clock::time_point end_time = steady_clock::now() + nanoseconds(timeout.toNSec());
+
     while (!info->finished_)
     {
-      info->finished_condition_.wait(lock);
+      if (timeout > ros::WallDuration(0.0))
+      {
+        if (info->finished_condition_.wait_until(lock, end_time) == boost::cv_status::timeout)
+        {
+          ROS_WARN("Service [%s] call failed: no response for %fsec", service_name_.c_str(), timeout.toSec());
+          timed_out = true;
+          break;
+        }
+      }
+      else
+      {
+        info->finished_condition_.wait(lock);
+      }
     }
   }
 
+  // Clear calls is properly protected
```

Comment:
I can't really see any deadlock on `info->finished_mutex_`.

However, it seems that on timeout we should do something with the `connection_`, so the read callback is cancelled somehow. Thoughts?

Another minor comment:
Why instead of this `if` block with `cancelCall`, don't we set `info->finished_ = true;` before the `break;` in the loop?
There we own the lock on `info->finished_mutex_`.

Reply:
> I can't really see any deadlock on info->finished_mutex_.

Sorry, the suspected deadlock was a wrong guess (because the test failed, but that was due to the "server" still processing the previous request, fixed in new tests....)

> Why instead of this if block with cancelCall, don't we set info->finished_ = true; before the break; in the loop?
> There we own the lock on info->finished_mutex_.

Tried that and turned out that is not working reliably (actually same for the cancelCall).